### PR TITLE
[Draft PR] LPS-186127 Prohibiting single/double-word language keys without [context]

### PR DESCRIPTION
**Pull request by ling-alan-huang**
Hi @brianwulbern 
I think we need to send the PR to you to let you know what we can catch with this SF task.
Let's improve it and make it fitable for the language keys together.

I kept the checking for all numbers in [], but we skip checking for it when the key contains `code` or `status`

You can try it on your end.
Steps:
1.` gradlew deploy` in `modules/util/source-formatter`
2. Set `source.check.PropertiesLanguageKeysContextCheck.enabled` to `true` to enable the SF check.
3. Set `source.check.PropertiesLanguageKeysContextCheck.contextDepth` for depth, it defaults to 0, means SF will check all keys without limitaion.
4. In `portal-impl` run ant SF(`format-source-all`, `format-source-current-branch`, `format-source-local-changes`), or even run ant format-source-files with the desired dir
For example: `ant format-source-files -Dsource.base.dir=/home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang`

Here is the result of `ant format-source-files -Dsource.base.dir=/home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang`
```
     [java] java.lang.Exception: Found 28 formatting issues:
     [java] 1: The key for 'add-category' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 2: The key for 'add-child-category' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 3: The key for 'categories' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 4: The key for 'category' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 5: The key for 'checkout' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 6: The key for 'delete' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 7: The key for 'enabled' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 8: The key for 'flag' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 9: The key for 'history' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 10: The key for 'is-temporarily-unavailable' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 11: The key for 'log-exceptions-only-help' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 12: The key for 'map' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 13: The key for 'mapping' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 14: The key for 'monitoring-configuration-name' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 15: The key for 'object' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 16: The key for 'parent-category' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 17: The key for 'pending' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 18: The key for 'recent' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 19: The key for 'redirect-url' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 20: The key for 'second' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 21: The key for 'subcategories' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 22: The key for 'then' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 23: The key for 'timeout' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 24: The key for 'title' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 25: The key for 'view' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 26: The key for 'vote' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 27: The key for 'when' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)
     [java] 28: The key for 'workflow-status' should have the context in [] to indicate the specific meaning: /home/alan/liferay_code/master/liferay-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)

```

### CODE REVIEW

**Comment by brianwulbern** [comment]
ok, cool - thanks for this @ling-alan-huang.  I'll check it out locally soon!

**Comment by brianwulbern** [comment]
@ling-alan-huang  hmm, I'm not sure what I'm missing.  I have `source.check.PropertiesLanguageKeysContextCheck.enabled=true` and `source.check.PropertiesLanguageKeysContextCheck.contextDepth=1` set in `source-formatter.properties`, then I ran `ant format-source-files -Dsource.base.dir=/home/liferay/project/liferay-portal/master-portal/modules/apps/portal-language/portal-language-lang` and it did this:
```
liferay@CN-FCXSZ33:~/project/liferay-portal/master-portal/portal-impl$ ant format-source-files -Dsource.base.dir=/home/liferay/project/liferay-portal/master-portal/modules/apps/portal-language/portal-language-lang
Buildfile: /home/liferay/project/liferay-portal/master-portal/portal-impl/build.xml
  [typedef] Could not load definitions from resource org/apache/maven/artifact/ant/antlib.xml. It could not be found.
  [taskdef] Could not load definitions from resource org/jacoco/ant/antlib.xml. It could not be found.
   [delete] Deleting: /home/liferay/project/liferay-portal/master-portal/null858302923.properties

format-source-files:
     [echo] Executing Gradle task: :portalYarnFormat
     [exec] Starting a Gradle Daemon (subsequent builds will be faster)
     [exec] Configuration on demand is an incubating feature.
     [exec] > Task :downloadNode
     [exec] > Task :npmInstall SKIPPED
     [exec] 
    [exec] > Task :yarnInstall
     [exec] yarn install v1.13.0
     [exec] warning Missing name in workspace at "/home/liferay/project/liferay-portal/master-portal/modules/dxp/apps/osb/osb-faro", ignoring.
     [exec] [1/4] Resolving packages...
     [exec] warning Resolution field "@types/html-minifier-terser@5.1.0" is incompatible with requested version "@types/html-minifier-terser@^6.0.0"
     [exec] [2/4] Fetching packages...
     [exec] info fsevents@1.2.7: The platform "linux" is incompatible with this module.
     [exec] info "fsevents@1.2.7" is an optional dependency and failed compatibility check. Excluding it from installation.
     [exec] info fsevents@2.3.1: The platform "linux" is incompatible with this module.
     [exec] info "fsevents@2.3.1" is an optional dependency and failed compatibility check. Excluding it from installation.
     [exec] [3/4] Linking dependencies...
     [exec] warning " > @liferay/eslint-plugin@1.4.0" has unmet peer dependency "eslint@>=4.1.1".
     [exec] warning "@liferay/eslint-plugin > @typescript-eslint/eslint-plugin@4.30.0" has unmet peer dependency "eslint@^5.0.0 || ^6.0.0 || ^7.0.0".
     [exec] warning "@liferay/eslint-plugin > @typescript-eslint/parser@4.30.0" has unmet peer dependency "eslint@^5.0.0 || ^6.0.0 || ^7.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-config-prettier@8.1.0" has unmet peer dependency "eslint@>=7.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-no-for-of-loops@1.0.0" has unmet peer dependency "eslint@>=2.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-notice@0.9.10" has unmet peer dependency "eslint@>=3.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-promise@5.1.1" has unmet peer dependency "eslint@^7.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-react@7.23.1" has unmet peer dependency "eslint@^3 || ^4 || ^5 || ^6 || ^7".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-react-hooks@4.2.0" has unmet peer dependency "eslint@^3.0.0 || ^4.0.0 || ^5.0.0 || ^6.0.0 || ^7.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-sort-destructure-keys@1.3.5" has unmet peer dependency "eslint@3 - 7".
     [exec] warning "@liferay/eslint-plugin > @typescript-eslint/eslint-plugin > @typescript-eslint/experimental-utils@4.30.0" has unmet peer dependency "eslint@*".
     [exec] warning "@liferay/eslint-plugin > @typescript-eslint/eslint-plugin > tsutils@3.21.0" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta"
```

**Comment by brianwulbern** [comment]
The second time it was faster, but didn't catch any formatting issues
```
format-source-files:
     [echo] Executing Gradle task: :portalYarnFormat
     [exec] Configuration on demand is an incubating feature.
     [exec] > Task :downloadNode UP-TO-DATE
     [exec] > Task :npmInstall SKIPPED
     [exec] 
     [exec] > Task :yarnInstall
     [exec] yarn install v1.13.0
     [exec] warning Missing name in workspace at "/home/liferay/project/liferay-portal/master-portal/modules/dxp/apps/osb/osb-faro", ignoring.
     [exec] [1/4] Resolving packages...
     [exec] warning Resolution field "@types/html-minifier-terser@5.1.0" is incompatible with requested version "@types/html-minifier-terser@^6.0.0"
     [exec] success Already up-to-date.
     [exec] Done in 1.20s.
     [exec] 
     [exec] > Task :packageRunFormat
     [exec] yarn run v1.13.0
     [exec] $ liferay-npm-scripts fix
     [exec] Done in 11.96s.
     [exec] 
     [exec] > Task :portalYarnFormat
     [exec] 
     [exec] > Task :portalWorkspacesYarnInstall
     [exec] yarn install v1.13.0
     [exec] [1/4] Resolving packages...
     [exec] success Already up-to-date.
     [exec] Done in 0.57s.
     [exec] 
     [exec] > Task :portalWorkspacesYarnFormat
     [exec] yarn run v1.13.0
     [exec] $ liferay-npm-scripts lint:fix && liferay-npm-scripts format
     [exec] Done in 0.94s.
     [exec] 
     [exec] BUILD SUCCESSFUL in 16s
     [exec] 5 actionable tasks: 4 executed, 1 up-to-date
     [exec] 

BUILD SUCCESSFUL
Total time: 40 seconds

```

**Comment by brianwulbern** [comment]
Ok, @ling-alan-huang, I ran it after getting the step 1 deploy to work.  And it doesn't seem to matter whether I put the property in my SF command, or in the properties file, I still get the same thing
```
ant format-source-files -Dsource.base.dir=/home/liferay/project/liferay-portal/master-portal/modules/apps/portal-language/portal-language-lang -Dsource.check.PropertiesLanguageKeysContextCheck.enabled=true
Buildfile: /home/liferay/project/liferay-portal/master-portal/portal-impl/build.xml
   [delete] Deleting: /home/liferay/project/liferay-portal/master-portal/null1512825989.properties

format-source-files:
     [java] Loading file:/home/liferay/project/liferay-portal/master-portal/portal-impl/classes/system.properties
     [echo] Executing Gradle task: :portalYarnFormat
     [exec] Configuration on demand is an incubating feature.
     [exec] > Task :downloadNode UP-TO-DATE
     [exec] > Task :npmInstall SKIPPED
     [exec] 
     [exec] > Task :yarnInstall
     [exec] yarn install v1.13.0
     [exec] warning Missing name in workspace at "/home/liferay/project/liferay-portal/master-portal/modules/dxp/apps/osb/osb-faro", ignoring.
     [exec] [1/4] Resolving packages...
     [exec] warning Resolution field "@types/html-minifier-terser@5.1.0" is incompatible with requested version "@types/html-minifier-terser@^6.0.0"
     [exec] success Already up-to-date.
     [exec] Done in 1.38s.
     [exec] 
     [exec] > Task :packageRunFormat
     [exec] yarn run v1.13.0
     [exec] $ liferay-npm-scripts fix
     [exec] Done in 13.55s.
     [exec] 
     [exec] > Task :portalYarnFormat
     [exec] 
     [exec] > Task :portalWorkspacesYarnInstall
     [exec] yarn install v1.13.0
     [exec] [1/4] Resolving packages...
     [exec] [2/4] Fetching packages...
     [exec] info fsevents@2.3.2: The platform "linux" is incompatible with this module.
     [exec] info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
     [exec] info fsevents@1.2.13: The platform "linux" is incompatible with this module.
     [exec] info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
     [exec] [3/4] Linking dependencies...
     [exec] warning " > @liferay/eslint-plugin@1.4.0" has unmet peer dependency "eslint@>=4.1.1".
     [exec] warning "@liferay/eslint-plugin > @typescript-eslint/eslint-plugin@4.30.0" has unmet peer dependency "eslint@^5.0.0 || ^6.0.0 || ^7.0.0".
     [exec] warning "@liferay/eslint-plugin > @typescript-eslint/parser@4.30.0" has unmet peer dependency "eslint@^5.0.0 || ^6.0.0 || ^7.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-config-prettier@8.1.0" has unmet peer dependency "eslint@>=7.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-no-for-of-loops@1.0.1" has unmet peer dependency "eslint@>=2.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-notice@0.9.10" has unmet peer dependency "eslint@>=3.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-promise@5.2.0" has unmet peer dependency "eslint@^7.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-react@7.23.1" has unmet peer dependency "eslint@^3 || ^4 || ^5 || ^6 || ^7".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-react-hooks@4.2.0" has unmet peer dependency "eslint@^3.0.0 || ^4.0.0 || ^5.0.0 || ^6.0.0 || ^7.0.0".
     [exec] warning "@liferay/eslint-plugin > eslint-plugin-sort-destructure-keys@1.5.0" has unmet peer dependency "eslint@3 - 8".
     [exec] warning "@liferay/eslint-plugin > @typescript-eslint/eslint-plugin > @typescript-eslint/experimental-utils@4.30.0" has unmet peer dependency "eslint@*".
     [exec] warning "@liferay/eslint-plugin > @typescript-eslint/eslint-plugin > tsutils@3.21.0" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta".
     [exec] warning "@liferay/eslint-plugin > @typescript-eslint/eslint-plugin > @typescript-eslint/experimental-utils > eslint-utils@3.0.0" has unmet peer dependency "eslint@>=5".
     [exec] warning "@liferay/npm-scripts > @testing-library/react@12.1.5" has incorrect peer dependency "react@<18.0.0".
     [exec] warning "@liferay/npm-scripts > @testing-library/react@12.1.5" has incorrect peer dependency "react-dom@<18.0.0".
     [exec] warning "@liferay/npm-scripts > liferay-lang-key-dev-loader@1.0.3" has incorrect peer dependency "webpack@^3.0.0 || ^4.0.0".
     [exec] warning "@liferay/npm-scripts > script-ext-html-webpack-plugin@2.1.4" has incorrect peer dependency "webpack@^1.0.0 || ^2.0.0 || ^3.0.0 || ^4.0.0".
     [exec] warning "@liferay/npm-scripts > script-ext-html-webpack-plugin@2.1.4" has incorrect peer dependency "html-webpack-plugin@^3.0.0 || ^4.0.0".
     [exec] [4/4] Building fresh packages...
     [exec] Done in 8.32s.
     [exec] 
     [exec] > Task :portalWorkspacesYarnFormat
     [exec] yarn run v1.13.0
     [exec] $ liferay-npm-scripts lint:fix && liferay-npm-scripts format
     [exec] Done in 1.65s.
     [exec] 
     [exec] BUILD SUCCESSFUL in 26s
     [exec] 5 actionable tasks: 4 executed, 1 up-to-date
     [exec] 
```

**Comment by brianwulbern** [comment]
> 3. Set source.check.PropertiesLanguageKeysContextCheck.contextDepth for depth, it defaults to 0, means SF will check all keys without limitaion.

Does that mean that if I set it to "1", it'll find keys with one word, "2" will find keys with up to two words?

**Comment by ling-alan-huang** [comment]
1 for a=
2 for a-b=

**Comment by ling-alan-huang** [comment]
Did you fetch this PR on your end?

**Comment by ling-alan-huang** [comment]
Rebased and added check for one word for both key and value, like `theme=Theme`

**Comment by brianwulbern** [comment]
Thanks @ling-alan-huang!  Although like Olaf [said](https://github.com/brianwulbern/liferay-portal/pull/426#issuecomment-1573412049), I think the check can be fine on just the key half of the equation.  It should be fine as you have it, though.

**Comment by brianwulbern** [comment]
Ok, I was able to generate an output of 3463 lines, but each single-word key seems to show up two times (with "[java] java.lang.Exception: Found ___ formatting issues:" in the middle), can we trim down the output a bit?

Current two mentions of the key (do we need it twice, or is that just the standard output style for SF?):

1. ```[java] One word 'about' for both key and value might have the disambiguation: /home/liferay/project/liferay-portal/master-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)```
2. ```[java] 3: One word 'about' for both key and value might have the disambiguation: /home/liferay/project/liferay-portal/master-portal/modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties (SourceCheck:PropertiesLanguageKeysContextCheck)```

There's also 26 instances of:
```
The key for '...' should have the context in [] to indicate the specific meaning
```

Can we have the output be similar to this ^ for **all** of them?  Something like:
```
The single-word key '...' should include at the end a word of context within [] to indicate specific meaning.
```

This one is a bit wordy:
```
One word '...' for both key and value might have the disambiguation
```

**Comment by ling-alan-huang** [comment]
Hey @brianwulbern 
Yes, that's the standard output style for SF, if there is no special needed I don't want to change it.

Also, for the messages I wanted to check with you before, can you provide the more proper messages for cases below?

_Case 1:_
**order=Order**
OUTPUT: `The key for 'order' should have the context in [] to indicate the specific meaning`
Use this: `The single-word key 'order' should include at the end a word of context within [] to indicate specific meaning`

_Case 2:_
**order[123]=Order**
OUTPUT: `The context '123' is invalid in the key 'order[123]'`

_Case 3:_
**theme=Theme**
OUTPUT: `One word 'theme' for both key and value might have the disambiguation`

**Comment by brianwulbern** [comment]
@ling-alan-huang  Ah, ok - I guess it only takes a ton of space if the scan returns thousands of results :sweat_smile: 

I think we can merge Case 3 into Case 1.  All of the single-word keys will need [context] at the end, indicating that which will help disambiguate.  I'll rearrange the wording order a bit, so it'll read more fluidly, and be like this:
`The single-word key 'order' should include a word of context at the end, within a [], to indicate specific meaning.`

Case 2 can stay for now, that's helpful validation, and the wording there is fine.

**Comment by ling-alan-huang** [comment]
@brianwulbern I used the message you provided only for single-word key case.

But I just simply dropped **"single-word"** from the message for this case:
```
redirect-url=Redirect URL
redirect-url[oauth]=OAuth Redirect URL
```
Because in this case the check for key is depended on `contextDepth`, if we set contextDepth=0, or 2
The output will be `The key 'redirect-url' should include a word of context at the end, within a [], to indicate specific meaning`

Let me know if we can improve the check.

**Comment by brianwulbern** [comment]
@ling-alan-huang ah, I see what you mean.  Yep, I like that change - good call :handshake:

**Comment by brianwulbern** [comment]
@ling-alan-huang  I'll fetch the latest code and run again to test.  Then I'll probably make changes to a couple keys and send it to BChan.

I'm thinking that in at least some cases it will be difficult to track down the language key being used in portal code (not to mention there being 1730 violations of this new rule), so if there are single-word keys that aren't ambiguous, what do we think of having a comma-delisted (+ new line) file to effectively 'whitelist' certain keys not lacking context?  Maybe it could also be something like `source.check.PropertiesLanguageKeysCheck.allowedLanguageKeys` used [here](https://github.com/liferay/liferay-portal/blob/master/source-formatter.properties#L771)

cc @olafk

**Comment by ling-alan-huang** [comment]
@brianwulbern Yes, sure thing, feel free to make any changes.

**Comment by ling-alan-huang** [comment]
@brianwulbern I rebased to latest mater and added additional logic for exclusions of single-word keys.

**Comment by brianwulbern** [comment]
see https://github.com/brianwulbern/liferay-portal/pull/429
